### PR TITLE
lexer, parser: Decorate errors with expression snippet.

### PR DESCRIPTION
**Pull request by qinqon**
At lexer, parser is easy to detect the rune, token that is generating
the error. This change wrap those errors with the expression and a
pointer.
    
Since the parser knows only about tokens, this change add also the
expression to the parser so it can use it to wrap the errors.

At resolver the ast.Node contains a Position field that points to the expressions
rune where the ast.Node was generated from. This change use that
position to to wrap the errors with a snippet that contains the
expression and a pointer to the place that generated the error.

To acomplish that the capture expressions are passed to the resolver so
it can search for the expression string to create the snippet.

The snippet pointer marks until operators but not inside the paths, this
can be fixed at a follow up.

The errors will look like:
```
invalid equality filter: right hand argument is not a string
| routes.running.destination==this.is.wrong                                     
| ............................^
```

### CODE REVIEW

**Comment by qinqon** [comment]
/hold
I am going to add some error to integration test to see how it looks

**Comment by qinqon** [comment]
/hold cancel

Integration tests for lexer/parser/resolver error has being added.

File: nmpolicy/internal/resolver/resolver.go
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver.go:114]: Let's not print `%+v` it is not readable for users. Let's add `String()` method to `ast.Node`. We can address this in a follow-up.

File: nmpolicy/internal/resolver/resolver.go
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver.go:178]: same here

File: nmpolicy/internal/resolver/resolver_test.go
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver_test.go:383]: Can you please use the real capture expression `invalid-path-type: "interfaces.ipv4.address==10.244.0.1"` instead of `"01234567"`?
For me it was hard to follow and see it the `^` place makes sense when using `"01234567"`.

File: tests/basic_test.go
Comment (AlonaKaplan) [inline tests/basic_test.go:470]: with parser error

File: tests/basic_test.go
Comment (AlonaKaplan) [inline tests/basic_test.go:495]: with resolver error

File: tests/basic_test.go
Comment (AlonaKaplan) [inline tests/basic_test.go:515]: in this specific case it would be more readable if the `^` would point to the `interfaces` or to the pipe. We can fix it in a follow-up.

File: nmpolicy/internal/resolver/resolver.go
Comment (qinqon) [reply nmpolicy/internal/resolver/resolver.go:114]: This should be addressed as part of fix on #44

File: nmpolicy/internal/resolver/resolver.go
Comment (qinqon) [reply nmpolicy/internal/resolver/resolver.go:178]: ditto

File: nmpolicy/internal/resolver/resolver_test.go
Comment (qinqon) [reply nmpolicy/internal/resolver/resolver_test.go:383]: Problem with that is that make more difficult to put correct pos at AST.

File: tests/basic_test.go
Comment (qinqon) [reply tests/basic_test.go:515]: We will fix it at follow up after merging the PR that use a visitor for path.go

File: nmpolicy/internal/resolver/resolver_test.go
Comment (AlonaKaplan) [reply nmpolicy/internal/resolver/resolver_test.go:383]: I understand, but I believe it worth the effort since it will make the test much more readable.

File: tests/basic_test.go
Comment (qinqon) [reply tests/basic_test.go:515]: Added the visitor part at the issue that covers the path.go refactoring https://github.com/nmstate/nmpolicy/issues/50.

**Comment by qinqon** [comment]
I have mix a wip on different PR here that's why integration tests is failing...

**Comment by qinqon** [comment]
@AlonaKaplan PR ready now, have removed the noise from other PR.

File: nmpolicy/internal/resolver/resolver_test.go
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver_test.go:99]: Why marshalling the pool is required if we unmarshal it anyway in the `runTest`?

Comment (qinqon) [reply nmpolicy/internal/resolver/resolver_test.go:99]: So we don't have to change the rest of the code, since it's a test is not that problematic.

Comment (AlonaKaplan) [reply nmpolicy/internal/resolver/resolver_test.go:99]: What rest of the code?
Do you just need to change in `runTest`-
`captureASTPool := typestest.ToCaptureASTPool(t, testToRun.captureASTPool)` to
`captureASTPool := testToRun.captureASTPool`

Comment (qinqon) [reply nmpolicy/internal/resolver/resolver_test.go:99]: then we have to change all the test that are using "captureASTPool" as a string.

**Review by AlonaKaplan — APPROVED** [review]

**Pull request merged**
